Tests wait until microserver is ready

d2r opened the pull request:
This change also updates some tests that use "observers" to account for the
extra request served by the microserver before actual tests run.

The purpose of this patch is to help eliminate a race condition when running tests, between the time a port is bound and the time a microserver is able to serve requests.

The cause was traced down from a transient test error in #3360 pointed out by @zwoop

d2r commented on tests/gold_tests/autest-site/microserver.test.ext:
```diff
@@ -33,22 +37,6 @@ def httpObject(self, header, data):
-    absFilepath = os.path.abspath(filename)
-    self.Setup.CopyAs(absFilepath, self.Variables.DataDir)
-    return
-
```
This method was overloaded with a form taking one fewer positional arguments, and it is the override below that is added to the process object.

As far as I can tell, this is unused.

d2r commented on tests/gold_tests/continuations/double.test.py:
```diff
@@ -32,7 +32,7 @@
 
 Test.testName = ""
-request_header = {"headers": "GET / HTTP/1.1\r\n\r\n", "timestamp": "1469733493.993", "body": ""}
+request_header = {"headers": "GET / HTTP/1.1\r\nHost: double.test\r\n\r\n", "timestamp": "1469733493.993", "body": ""}
```
This request header is [rejected](https://github.com/apache/trafficserver/blob/8d01164257417434b667021365b5f07d0ec596e0/tests/tools/sessionvalidation/sessionvalidation.py#L200-L203) by the microserver, resulting in other responses being discarded. So I had to fix this so that the health check could function.

d2r commented on tests/gold_tests/headers/forwarded.gold:
```diff
@@ -2,6 +2,8 @@ FORWARDED MISSING
 FORWARDED MISSING
 -
+FORWARDED MISSING
+-
```
This output is from an observer, which begins observing even before the tests run. The extra entry here is due to the health check.

SolidWallOfCode replied:
Yes, this was my concern about the health checks. The observer is called from the micro-server on every request, *including the health check* request. I think we will need to tweak the micro-server to be able to mark a request/response as a health check in which case it does not call the observers.

d2r commented on tests/gold_tests/headers/via.gold:
```diff
@@ -1,3 +1,4 @@
+Via: ---missing---
```
Again, the observer begins recording before tests run, so it observes the health check happening before testing begins.

d2r commented on tests/gold_tests/transaction/txn.test.py:
```diff
@@ -33,7 +33,7 @@
 
 Test.testName = ""
-request_header = {"headers": "GET / HTTP/1.1\r\n\r\n",
+request_header = {"headers": "GET / HTTP/1.1\r\nHost: txn.test\r\n\r\n",
```
Missing host headers are [rejected](https://github.com/apache/trafficserver/blob/8d01164257417434b667021365b5f07d0ec596e0/tests/tools/sessionvalidation/sessionvalidation.py#L195-L198) when the microserver loads its sessions, and so the health check fails unless this is corrected.

d2r commented on tests/gold_tests/transaction/txn.test.py:
```diff
@@ -52,15 +52,15 @@
 
 ts.Disk.remap_config.AddLine(
-    'map http://127.0.0.1:{0} http://127.0.0.1:{1}'.format(
+    'map http://txn.test:{0} http://127.0.0.1:{1}'.format(
```
Another remap rule [rejected](https://github.com/apache/trafficserver/blob/8d01164257417434b667021365b5f07d0ec596e0/tests/tools/sessionvalidation/sessionvalidation.py#L200-L203) by the microserver, preventing the health check.

d2r commented on tests/tools/sessionvalidation/sessionvalidation.py:
```diff
@@ -197,11 +197,6 @@ def validateSingleTransaction(txn):
-            if "127.0.0.1" in txn_req.getHeaders() or "localhost" in txn_req.getHeaders():
-                _verbose_print("transaction request Host is localhost, we must reject because ATS ignores remap rules for localhost requests")
-                retval = False
-
```
This does not seem to have value, and it gets in the way of health checks that connect to the microserver at localhost directly.

zwoop commented:
Not sure what all the comments are about. @d2r are you complaining about your own PR? :-) Or should we land this?

d2r commented:
I added comments to explain the changes, since many of them might seem
out-of-scope.


-- 
Derek

SolidWallOfCode commented on tests/gold_tests/autest-site/microserver.test.ext:
```diff
@@ -144,6 +131,39 @@ def makeHeader(self, requestString, **kwargs):
 
+def uServerUpAndRunning(host, port, isSsl):
+    plain_sock = socket.socket(socket.AF_INET)
+    sock = ssl.wrap_socket(plain_sock) if isSsl else plain_sock
```
Why not `plain_sock = ssl.wrap_socket(plain_sock) if isSsl;`?

d2r replied:
Sure I can make this change.

d2r replied:
* Actually, python does not allow this kind of ternary syntax. You need an else.
* If the socket is wrapped than it is no longer "plain." 

I think maybe I should leave it?

SolidWallOfCode replied:
Yeah, if Python requires the `else` you can't really improve it.

dragon512 replied:
I am not sure what you are saying here. test ? a : b in C++ is A if test else B
so 
`sock = ssl.wrap_socket(plain_sock) if isSsl else socket.socket(socket.AF_INET)
`

d2r replied:
I am not sure it would make sense to create a second socket in addition to `plain_sock` in the case `isSsl` is not true.

But I'm willing to change to whatever style folks want.

PSUdaemon commented:
@zwoop @d2r Personally, I like this style of review. Answers a lot of basic questions a reviewer might have. The other way to do this would be several commits, but then you have to squash afterwards and regain approvals to merge. This seems fine to me.

d2r commented:
> I think we will need to tweak the micro-server to be able to mark a request/response as a health check in which case it does not call the observers.

Discussed with @SolidWallOfCode, I think we can do this. Not only is it a cleaner change, but it may leave the `uWServer.py` a bit cleaner as well. Let me look at it and post here when done.

d2r commented:
> The other way to do this would be several commits, but then you have to squash afterwards and regain approvals to merge.

I have been strongly advised—in this project—to keep my PR branches squashed to the minimum number of commits that make sense, because it has happened in the past that PRs get merged accidentally without having squashed commits.

On the other hand, as a reviewer I find it frustrating when someone squashes their commits while addressing review comments. Unless I keep their old history, I cannot easily see what has changed since I last looked.

Given that, I am happy to do what the release manager wants, since managing release branches is a choke point where messes must eventually be handled.

persiaAziz commented on tests/gold_tests/autest-site/microserver.test.ext:
```diff
@@ -160,16 +180,27 @@ def MakeOriginServer(obj, name, port=False, ip=False, delay=False, ssl=False, lo
+        "body": "imok"
+    })
+
+    p.Ready = When.uServerUpAndRunning(ip, port, ssl)
```
Do we have any case when this may return false?

d2r replied:
The `p.Ready =` statement stores the method pointed to for future execution, so the rvalue is just the method.

The `uServerUpAndRunning`, when executed, should either return `True` (check passed), raise an exception (unexpected response), or just keep trying with a sleep interval of `.1` seconds. As I understand it, there is a timeout that is watched somewhere in the caller chain that will "fail" (error) the test if reached before the check returns.

d2r replied:
Actually, when I tested this assumption, I found that this seems to check for a timeout in the same thread, so I need to return `False` rather than loop on an interval. If something strange prevents the server from listening for TCP connections, then we would never error out. I'll fix that.

d2r commented on tests/tools/microServer/uWServer.py:
```diff
@@ -36,7 +36,7 @@
 test_mode_enabled = True
 lookup_key_ = "{PATH}"
-__version__ = "1.0"
+__version__ = "1.1"
```
We are adding a new key to the session definition, so I thought it would be good to bump the minor version.

Are there any other consequences of doing this?

d2r replied:
Following up: This is the version of the command-line tool that is printed and not a schema version. So it is safe to change it.

d2r commented:
Review comments should be addressed.

d2r commented:
Please hold off on merging until I investigate missing documentation...

d2r commented:
```Diff
diff --git a/tests/gold_tests/autest-site/microserver.test.ext b/tests/gold_tests/autest-site/microserver.test.ext
index 196faf8b1..a52e9b80e 100644
--- a/tests/gold_tests/autest-site/microserver.test.ext
+++ b/tests/gold_tests/autest-site/microserver.test.ext
@@ -111,7 +111,7 @@ def addTransactionToSession(txn, JFile):
 
     if jsondata == None:
         jsondata = dict()
-        jsondata["version"] = '0.1'
+        jsondata["version"] = '0.2'
         jsondata["timestamp"] = "1234567890.098"
         jsondata["encoding"] = "url_encoded"
         jsondata["txns"] = list()
@@ -185,7 +185,7 @@ def MakeOriginServer(obj, name, port=False, ip=False, delay=False, ssl=False, lo
     AddMethodToInstance(p, addTransactionToSession)
 
     # Set up health check.
-    addResponse(p, "sessionfile.log", {
+    addResponse(p, "healthcheck.json", {
         "headers": "GET /ruok HTTP/1.1\r\nHost: {}\r\n\r\n".format(ip),
         "timestamp": "1469733493.993",
         "body": ""
diff --git a/tests/tools/microServer/README.md b/tests/tools/microServer/README.md
index 0a8fa6535..a20ea85dc 100644
--- a/tests/tools/microServer/README.md
+++ b/tests/tools/microServer/README.md
@@ -4,20 +4,7 @@ uWServer
 uWServer is a mock HTTP server that takes predefined set of sessions for serving response to HTTP requests. Each session includes one or more transactions. A transaction is composed of an HTTP request and an HTTP response. 
 uWServer accepts session data in JSON fromat only.
 
-Example session :
-```
- {"version": "0.1",  
-  "txns": [  
-        {"request": {"headers": "GET /path1\r\n Host: example.com \r\n\r\n", "timestamp": "1522783378", "body": "Apache Traffic Server"},     
-        "response": {"headers": "HTTP/1.1\r\n Server: microserver\r\nContent-Length:100 \r\n\r\n", "timestamp": "1522783378", "body": ""},  
-         "uuid": "1"},   
-        {"request": {"headers": "GET /path2\r\n\r\n", "timestamp": "1522783378", "body": "Apache Traffic Server"},   
-        "response": {"headers": "HTTP/1.1\r\nTransfer-Encoding: chunked\r\n\r\n", "timestamp": "1522783378", "body": "Apache Traffic Server"},   
-        "uuid": "2"}  
-  ],   
-  "timestamp": "1522783378",   
-  "encoding": ""}  
-```
+
 Command:
 ----------------
 
@@ -26,6 +13,37 @@ Command:
 Options:
 -----------
 
-To see the options please run `python3.5 uWServer.py -h`
+To see the options please run `python3.5 uWServer.py --help`
+
+Session Definitions:
+--------------------
+
+Example session:
+
+```
+{
+  "encoding": "url_encoded",
+  "version": "0.2",
+  "txns": [
+    {
+      "response": {
+        "headers": "HTTP/1.1 200 OK\r\nConnection: close\r\n\r\n",
+        "body": "",
+        "timestamp": "1469733493.993"
+      },
+      "request": {
+        "headers": "GET / HTTP/1.1\r\nHost: www.example.test\r\n\r\n",
+        "body": "",
+        "timestamp": "1469733493.993"
+      },
+      "uuid": "",
+      "timestamp": ""
+    }
+  ],
+  "timestamp": "1234567890.098"
+}
+```
 
+Each session should be in its own file, and any number of files may be created to define sessions.
 
+The `response` map may include an `options` string, which is a comma-delimited list of options to be enabled. Currently the only option supported is `skipHooks`, which will ignore any hooks created for the matching request/response pair. See **Options**.
```
Changes since last review are in the above diff. They are as follows:

* @persiaAziz has added some documentation. To this I added info about the options, and re-arranged the example.
* I bumped the session definition versions we use, since we have added options.
* I renamed the JSON file we used to define the health check. This should make the purpose of the session clear to someone debugging a test.

d2r commented:
@persiaAziz or @SolidWallOfCode how does this look?

dragon512 commented:
[approve ci autest]
